### ISSUE TITLE
[Linux] "A KeyUpEvent is dispatched, but the state shows that the physical key is pressed on a different logical key" error 

### ISSUE THREAD

**Issue by MatthewLM** [flutter/flutter#89982]
Keyboard input is broken on Linux when typing "@". This problem did not occur when using flutter `2.2.3`. It was introduced upon upgrading to `2.5.0`.

## Steps to Reproduce

1. Run `flutter create bug`.
2. Update the files as follows: Just add a `TextFormField()`.
3. Type "hello@hello.com" into the `TextFormField`.

**Expected results:**

I expected to see "hello@hello.com"

**Actual results:**

Only "hello@" is shown and text input no longer works.

<details>
  <summary>Logs</summary>

```
[+146188 ms] ** (bug:31682): WARNING **: 15:30:43.554: Unable to retrieve framework response: Message is not valid JSON
[   +6 ms]
           ══╡ EXCEPTION CAUGHT BY SERVICES LIBRARY ╞══════════════════════════════════════════════════════════
           The following assertion was thrown during a platform message callback:
           A KeyUpEvent is dispatched, but the state shows that the physical key is pressed on a different
           logical key. If this occurs in real application, please report this bug to Flutter. If this occurs
           in unit tests, please ensure that simulated events follow Flutter's event model as documented in
           `HardwareKeyboard`. This was the event: KeyUpEvent#de59f(physicalKey:
           PhysicalKeyboardKey#70034(usbHidUsage: "0x00070034", debugName: "Quote"), logicalKey:
           LogicalKeyboardKey#00027(keyId: "0x00000027", keyLabel: "'", debugName: "Quote Single"), character:
           null, timeStamp: 5:16:04.135000) and the recorded logical key LogicalKeyboardKey#00040(keyId:
           "0x00000040", keyLabel: "@", debugName: "At")
           'package:flutter/src/services/hardware_keyboard.dart':
           Failed assertion: line 444 pos 16: '_pressedKeys[event.physicalKey] == event.logicalKey'

           Either the assertion indicates an error in the framework itself, or we should provide substantially
           more information in this error message to help you determine and fix the underlying cause.
           In either case, please report this assertion by filing a bug on GitHub:
             https://github.com/flutter/flutter/issues/new?template=2_bug.md

           When the exception was thrown, this was the stack:
           #2      HardwareKeyboard._assertEventIsRegular.<anonymous closure> (package:flutter/src/services/hardware_keyboard.dart:444:16)
           #3      HardwareKeyboard._assertEventIsRegular (package:flutter/src/services/hardware_keyboard.dart:452:6)
           #4      HardwareKeyboard.handleKeyEvent (package:flutter/src/services/hardware_keyboard.dart:543:5)
           #5      KeyEventManager.handleRawKeyMessage (package:flutter/src/services/hardware_keyboard.dart:821:35)
           #6      BasicMessageChannel.setMessageHandler.<anonymous closure> (package:flutter/src/services/platform_channel.dart:73:49)
           #7      BasicMessageChannel.setMessageHandler.<anonymous closure> (package:flutter/src/services/platform_channel.dart:72:47)
           #8      _DefaultBinaryMessenger.setMessageHandler.<anonymous closure> (package:flutter/src/services/binding.dart:379:35)
           #9      _DefaultBinaryMessenger.setMessageHandler.<anonymous closure> (package:flutter/src/services/binding.dart:376:46)
           #10     _invoke2.<anonymous closure> (dart:ui/hooks.dart:205:15)
           #14     _invoke2 (dart:ui/hooks.dart:204:10)
           #15     _ChannelCallbackRecord.invoke (dart:ui/channel_buffers.dart:42:5)
           #16     _Channel.push (dart:ui/channel_buffers.dart:132:31)
           #17     ChannelBuffers.push (dart:ui/channel_buffers.dart:329:17)
           #18     PlatformDispatcher._dispatchPlatformMessage (dart:ui/platform_dispatcher.dart:544:22)
           #19     _dispatchPlatformMessage (dart:ui/hooks.dart:92:31)
           (elided 5 frames from class _AssertionError and dart:async)
           ════════════════════════════════════════════════════════════════════════════════════════════════════

```

```
Analyzing bug...
No issues found! (ran in 1.1s)
```

Relevant parts:
```
[✓] Flutter (Channel stable, 2.5.0, on Debian GNU/Linux 10 (buster) 4.19.0-17-amd64, locale en_GB.UTF-8)
    • Flutter version 2.5.0...
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 4cc385b4b8 (5 days ago), 2021-09-07 23:01:49 -0700
    • Engine revision f0826da7ef
    • Dart version 2.14.0

[✓] Linux toolchain - develop for Linux desktop
    • clang version 7.0.1-8+deb10u2 (tags/RELEASE_701/final)
    • cmake version 3.13.4
    • ninja version 1.8.2
    • pkg-config version 0.29

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Debian GNU/Linux 10 (buster) 4.19.0-17-amd64
    • Chrome (web)    • chrome • web-javascript • Chromium 90.0.4430.212 built on Debian 10.9, running on Debian 10.10
```

</details>

**Comment by narumishi** [comment]
Same here macOS. Even we don't input text in textfield. Just press and hold `cmd` or other keys and move focus between flutter app and another program, sometimes(not always) the error will be raised.

**Comment by maheshmnj** [comment]
> Same here macOS

@narumishi  please file a separate issue with exact steps and logs, This issue seems a bit different from what you are reporting.

**Comment by devtud** [comment]
This exception is also raised when **Right Alt** is pressed and, as I mentioned in [this issue](https://github.com/flutter/flutter/issues/89981) and when you combine **Shift** with any other **number key** (not only 2 for @), **question mark** and _possibly_ others.

**Comment by whesse** [comment]
A similar exception is being thrown on Flutter web, on Chrome 93 on Linux, with Flutter 2.5.0.

In this case, after a minute of typing and changing focus in the web app, every keystroke generates an exception on the console, but the key also has its expected effect when an input field has focus.

The exception is "A KeyDownEvent is dispatched, but the state shows that the physical key is already pressed."  The KeyDownEvent physical key and logical key are the "Arrow Right" key, even though I am pressing a completely different key.

I am trying to find a reproducible sequence of key presses, copy-pastes, and focus changes that cause this error to start, but haven't found it yet.  I can consistently enter it after about a minute of hitting keys.

The web app is "current_results_ui" at https://github.com/dart-lang/dart_ci/tree/master/current_results_ui.
I will continue verifying the problem, by testing with earlier versions of Flutter web and finding a reproducible test case.

The error stack is

````
Exception caught by services library 
The following assertion was thrown during a platform message callback:
Assertion failed: file:///src/flutter/packages/flutter/lib/src/services/hardware_keyboard.dart:437:16
!_pressedKeys.containsKey(event.physicalKey)
"A KeyDownEvent is dispatched, but the state shows that the physical key is already pressed. If this occurs in real application, please report this bug to Flutter. If this occurs in unit tests, please ensure that simulated events follow Flutter's event model as documented in `HardwareKeyboard`. This was the event: KeyDownEvent#8e77c(physicalKey: PhysicalKeyboardKey#7004f(usbHidUsage: \"0x0007004f\", debugName: \"Arrow Right\"), logicalKey: LogicalKeyboardKey#55c9d(keyId: \"0x100000303\", keyLabel: \"Arrow Right\", debugName: \"Arrow Right\"), character: null, timeStamp: 0:00:51.308000)"

When the exception was thrown, this was the stack: 
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 251:49  throw_
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 29:3    assertFailed
packages/flutter/src/services/hardware_keyboard.dart 437:47                   <fn>
packages/flutter/src/services/hardware_keyboard.dart 451:14                   [_assertEventIsRegular]
packages/flutter/src/services/hardware_keyboard.dart 543:5                    handleKeyEvent
packages/flutter/src/services/hardware_keyboard.dart 821:35                   handleRawKeyMessage
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 84:54            runBody
dart-sdk/lib/_internal/js_dev_runtime/patch/async_patch.dart 123:5            _async
packages/flutter/src/services/hardware_keyboard.dart 808:51                   handleRawKeyMessage
...
````

**Comment by markusaksli-nc** [comment]
I wasn't able to reproduce this by just writing `hello@` so I randomly typed for a while and found that left shift + left alt + any other alphanumeric key will cause all input to break and start throwing these
```
** (master:15753): WARNING **: 16:10:33.322: Unable to retrieve framework response: Message is not valid JSON

======== Exception caught by services library ======================================================
The following assertion was thrown during a platform message callback:
A KeyUpEvent is dispatched, but the state shows that the physical key is pressed on a different logical key. If this occurs in real application, please report this bug to Flutter. If this occurs in unit tests, please ensure that simulated events follow Flutter's event model as documented in `HardwareKeyboard`. This was the event: KeyUpEvent#e93cb(physicalKey: PhysicalKeyboardKey#700e2(usbHidUsage: "0x000700e2", debugName: "Alt Left"), logicalKey: LogicalKeyboardKey#00106(keyId: "0x200000106", keyLabel: "Meta Left", debugName: "Meta Left"), character: null, timeStamp: 0:17:08.056000) and the recorded logical key LogicalKeyboardKey#00104(keyId: "0x200000104", keyLabel: "Alt Left", debugName: "Alt Left")
'package:flutter/src/services/hardware_keyboard.dart':
Failed assertion: line 444 pos 16: '_pressedKeys[event.physicalKey] == event.logicalKey'


Either the assertion indicates an error in the framework itself, or we should provide substantially more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
  https://github.com/flutter/flutter/issues/new?template=2_bug.md

When the exception was thrown, this was the stack: 
#2      HardwareKeyboard._assertEventIsRegular.<anonymous closure> (package:flutter/src/services/hardware_keyboard.dart:444:16)
#3      HardwareKeyboard._assertEventIsRegular (package:flutter/src/services/hardware_keyboard.dart:452:6)
#4      HardwareKeyboard.handleKeyEvent (package:flutter/src/services/hardware_keyboard.dart:543:5)
#5      KeyEventManager.handleRawKeyMessage (package:flutter/src/services/hardware_keyboard.dart:821:35)
#6      BasicMessageChannel.setMessageHandler.<anonymous closure> (package:flutter/src/services/platform_channel.dart:77:49)
#7      BasicMessageChannel.setMessageHandler.<anonymous closure> (package:flutter/src/services/platform_channel.dart:76:47)
#8      _DefaultBinaryMessenger.setMessageHandler.<anonymous closure> (package:flutter/src/services/binding.dart:387:35)
#9      _DefaultBinaryMessenger.setMessageHandler.<anonymous closure> (package:flutter/src/services/binding.dart:384:46)
#10     _invoke2.<anonymous closure> (dart:ui/hooks.dart:205:15)
#14     _invoke2 (dart:ui/hooks.dart:204:10)
#15     _ChannelCallbackRecord.invoke (dart:ui/channel_buffers.dart:42:5)
#16     _Channel.push (dart:ui/channel_buffers.dart:132:31)
#17     ChannelBuffers.push (dart:ui/channel_buffers.dart:329:17)
#18     PlatformDispatcher._dispatchPlatformMessage (dart:ui/platform_dispatcher.dart:544:22)
#19     _dispatchPlatformMessage (dart:ui/hooks.dart:92:31)
(elided 5 frames from class _AssertionError and dart:async)
====================================================================================================
```
Left Alt on its own will already cause this message before input breaks.
```
** (master:24714): CRITICAL **: 16:13:58.419: void synchronize_pressed_states_loop_body(gpointer, gpointer, gpointer): assertion 'pressed_logical_key_before_event == 0 || pressed_logical_key_before_event == logical_key' failed
```
This was known before as https://github.com/flutter/flutter/issues/89729 but this can happen without hot restarting once. I was able to reproduce this on the latest master `2.6.0-6.0.pre.76` as well as stable `2.5.0`
<details>
 <summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 2.6.0-6.0.pre.76, on Ubuntu 20.04.3 LTS 5.11.0-34-generic, locale en_US.UTF-8)
    • Flutter version 2.6.0-6.0.pre.76 at /home/dad/Development/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision fad5e4cf89 (17 hours ago), 2021-09-13 13:37:07 -0700
    • Engine revision abb1980f65
    • Dart version 2.15.0 (build 2.15.0-96.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /home/dad/Android/Sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /home/dad/.local/share/JetBrains/Toolbox/apps/AndroidStudio/ch-0/203.7678000/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 10.0.0-4ubuntu1
    • cmake version 3.16.3
    • ninja version 1.10.0
    • pkg-config version 0.29.1

[✓] Android Studio (version 2020.3)
    • Android Studio at /home/dad/.local/share/JetBrains/Toolbox/apps/AndroidStudio/ch-0/203.7678000
    • Flutter plugin version 60.1.2
    • Dart plugin version 203.8292
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Ubuntu 20.04.3 LTS 5.11.0-34-generic
    • Chrome (web)    • chrome • web-javascript • Google Chrome 93.0.4577.82

• No issues found!
```
</details>

**Comment by markusaksli-nc** [comment]
Doesn't reproduce on `2.2.3` so its a regression
<details>
 <summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel unknown, 2.2.3, on Linux, locale en_US.UTF-8)
    • Flutter version 2.2.3 at /home/dad/Development/flutter_old
    • Framework revision f4abaa0735 (3 months ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /home/dad/Android/Sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /home/dad/.local/share/JetBrains/Toolbox/apps/AndroidStudio/ch-0/203.7678000/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 10.0.0-4ubuntu1
    • cmake version 3.16.3
    • ninja version 1.10.0
    • pkg-config version 0.29.1

[✓] Android Studio (version 2020.3)
    • Android Studio at /home/dad/.local/share/JetBrains/Toolbox/apps/AndroidStudio/ch-0/203.7678000
    • Flutter plugin version 60.1.2
    • Dart plugin version 203.8292
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Linux
    • Chrome (web)    • chrome • web-javascript • Google Chrome 93.0.4577.82

• No issues found!

```
</details>

**Comment by markusaksli-nc** [comment]
Confirming what @whesse said it also reproduces on Chrome on Linux host. Caused by the same thing?

**Comment by whesse** [comment]
TLDR: The flutter web failure is due to fragile key handling code that fails if an exception is thrown by scrollable.position during handling of an arrow key.  The arrow key is registered as pressed, but never reset to unpressed because the exception happened.

The solution is not to have broken ScrollControllers, by not using the default primary ScrollController for multiple widgets on the page.

The Flutter web exception is being thrown because an error in the application is causing the ScrollController to throw an exception in the middle of the key handling code.  Because the application is using the default primary ScrollController for many Scrollable objects, that controller throws an exception when it is referenced, as in the "position" call made by a key press that might be caught by a scrollable, like arrow left, right, up, or down.  The key handling code then has set the "pressed" values to include that arrow key, but exits with the scrollable exception before it processes the key up and removes the "pressed" state of that key.

So this issue report can be ignored, unless the opinion is that key handling should be robust when exceptions are thrown by the scroll controller.







The flutter web KeyDown exception for right arrow reproduces reliably on flutter dev channel  2.6.0-5.1.pre  if right arrow is the first key pressed when the web page is opened.

The error does not reproduce on Flutter version 2.3.0-24.1.pre

**Comment by MatthewLM** [comment]
I should mention that I have a UK keyboard layout. `'` is the non-shifted character for the same key.

**Comment by darshankawar** [comment]
This seems to be occuring on Windows too, based on this https://github.com/flutter/flutter/issues/90050. Adding `platform-windows` label as applicable.

**Comment by CripyIce** [comment]
Same happens to me running MacOS with iOS emulator.. Can't press any key!
Only sometimes I can see a letter has been wrote in the TextField.

**Comment by cbracken** [comment]
Bumping priority given how serious this is.

**Comment by larryaasen** [comment]
I have the same issue on iOS simulator using Channel dev, 2.6.0-5.1.pre.

<details>
  <summary>Doctor</summary>

```
[✓] Flutter (Channel dev, 2.6.0-5.1.pre, on macOS 11.4 20F71 darwin-x64, locale en-US)
    • Flutter version 2.6.0-5.1.pre at /Applications/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f2903d3da5 (2 weeks ago), 2021-09-08 12:44:29 -0700
    • Engine revision 0bd9be379a
    • Dart version 2.15.0 (build 2.15.0-82.0.dev)

[!] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/larry/Library/Android/sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.

[✓] Xcode - develop for iOS and macOS (Xcode 12.5.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 50.0.1
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.60.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.23.0

[✓] Connected device (4 available)
    • iPhone 11 (mobile) • 149F69C8-D689-47AC-AA39-1A224BA1B090 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-5
      (simulator)
    • iPhone 12 (mobile) • 25E330C8-D67C-41FD-A04F-DD696084B2A0 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-5
      (simulator)
    • macOS (desktop)    • macos                                • darwin-x64     • macOS 11.4 20F71 darwin-x64
    • Chrome (web)       • chrome                               • web-javascript • Google Chrome 93.0.4577.82
    ! Error: Apple TV is not connected. Xcode will continue when Apple TV is connected. (code -13)
```

</details>

**Comment by mirbostani** [comment]
I have this issue on Linux using `Flutter 2.6.0-12.0.pre.85 • channel master`. It occurs while I press `shift + 8` on a `TextField` instance; however, only the left shift button raises the exception:

**Comment by doughywilson** [comment]
Also having this issue on Linux desktop builds.

**Comment by ibrahimyu** [comment]
Any clue that might cause this, so maybe we can find a workaround or even a possible fix?

**Comment by dkwingsmt** [comment]
This thread contains multiple issues that are possibly unrelated. They show the same error because they are triggered by the same assertion that checks the regularity of the incoming event sequence, the failure of which can come from various sources.

I'm restraining this thread exclusively for Linux for clarity. I'm looking at each issue one by one, but I can't reproduce all of the issues. 

When reporting issues, please make sure to add your `flutter doctor -v`, keyboard layout, and the reproduction steps. Thank you.

**Comment by KnucklesEQ** [comment]
I think this is the same issue discussed in #68713.

The problem with this issue is that usually Flutter just displays the error in the terminal... but sometimes, the TextField stops working.

**Comment by KyleKun** [comment]
Any expectations when this is getting fixed? Just asking because I am making a game for GitHub's Game Off (submission ends on December 1st) and this issue is a total deal breaker since the game suddenly crashes on desktop 😢

**Comment by dkwingsmt** [comment]
@mirbostani Thank you! I can reproduce it on Linux. I'm taking a look.

**Comment by dkwingsmt** [comment]
A fix has been merged and landed onto master.  https://github.com/flutter/engine/pull/29550
It should resolve the problem raised by @mirbostani.

Can everyone try to see if your problem persists?

**Comment by dkwingsmt** [comment]
@KyleKun That is not a desired move... I just saw the list of issues that were closed in favor of this one. I'll take a look at them one by one.

**Comment by dkwingsmt** [comment]
I've just gone over the list of issues that were "marked as duplicates" above. Some of them have been confirmed fixed. Some I can't reproduce, but I've unlocked and requested to verify on the latest master. 
Thank you for all the reports.

**Comment by dkwingsmt** [comment]
I'm marking this issue as fixed. If the same or similar issues occur, feel free to comment, reopen, or file a new issue.

(Yes I'm aware that other issue are in the wild, namely:
- https://github.com/flutter/flutter/issues/68713
- https://github.com/flutter/flutter/issues/93278
